components/logins/src/encryption.rs

```diff
@@ -26,24 +26,140 @@
 // multiple layers, from the app saying "sync now" all the way down to the
 // low level sync code.
 // To make life a little easier, we do that via a struct.
+//
+// Consumers of the Login component have 3 options for setting up encryption:
+//    1. Implement EncryptorDecryptor directly
+//       eg `LoginStore::new(MyEncryptorDecryptor)`
+//    2. Implement KeyManager and use ManagedEncryptorDecryptor
+//       eg `LoginStore::new(ManagedEncryptorDecryptor::new(MyKeyManager))`
+//    3. Generate a single key and create a StaticKeyManager and use it together with
+//       ManagedEncryptorDecryptor
+//       eg `LoginStore::new(ManagedEncryptorDecryptor::new(StaticKeyManager { key: myKey }))`
 
 use crate::error::*;
+use std::sync::Arc;
+
+/// This is the generic EncryptorDecryptor trait, as handed over to the Store during initialization.
+/// Consumers can implement either this generic trait and bring in their own crypto, or leverage the
+/// ManagedEncryptorDecryptor below, which provides encryption algorithms out of the box.
+///
+/// Note that EncryptorDecryptor must not call the db mutex of the LoginStore to prevent a
```

Comment:
I don't quite understand why this comment is relevant here - *nothing* probably should be taking that mutex other than the store (and, sadly, the sync engine)? If not for the sync engine that mutex could be private - there's probably ways to fix that which might be worthwhile in the future?

(I don't really care about the comment being here, I'm just trying to make sure I understand what's going on :)

Reply:
@bendk has addressed this problem – I think that if a consumer were to implement it (we all know that this shouldn't be the case, but hey, what's possible will be implemented at some point) and use the (pub) `db` property of the `LoginStore` in the `EncyptorDecryptor` implementation, then this could result in a deadlock. Did I understand that right, @bendk? And how can we rephrase the comment to be more precise?

Reply:
I raised the point since I think there's a subtle risk here.  Maybe the comment could focus more what EncryptorDecryptor implementers should not do and make the part about the mutex secondary.

Note that EncryptorDecryptor must not call any LoginStore methods.  The login store can call out to the EncryptorDecryptor when it's internal mutex is held so calling back in to the LoginStore may deadlock.